austin/format/pstat.py

```diff
@@ -0,0 +1,138 @@
+# This file is part of "austin-python" which is released under GPL.
+#
+# See file LICENCE or go to http://www.gnu.org/licenses/ for full license
+# details.
+#
+# austin-python is a Python wrapper around Austin, the CPython frame stack
+# sampler.
+#
+# Copyright (c) 2018-2020 Gabriele N. Tornetta <phoenix1987@gmail.com>.
+# All rights reserved.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from pstats import Stats, add_func_stats
+from austin.stats import Sample, InvalidSample
+
+ProfileName = str
+
+
+class Pstats:
+    def __init__(self):
+        self.last_sample = 0
+        self.last_stack = []
+        self.stats = Stats()
+
+    def print_stats(self):
+        self.stats.print_stats()
+
+    def dump(self, f):
+        self.stats.dump_stats(f)
+
+    def add_sample(self, sample: Sample):
+        # Stats.stats is a dictionary with key:
+        #  (fcode.co_filename, fcode.co_firstlineno, fcode.co_name)
+        # and value  (cc, ns, tt, ct, callers)
+        #     [cc] = The number of times this function was called, not counting direct
+        #           or indirect recursion,
+        #     [ns] = Number of times this function appears on the stack, minus one
+        #     [tt] = Total time spent internal to this function
+        #     [ct] = Cumulative time that this function was present on the stack.  In
+        #           non-recursive functions, this is the total execution time from start
+        #           to finish of each invocation of a function, including time spent in
+        #           all subfunctions.
+        #     [callers] = A dictionary indicating for each function name, the number of times
+        #           it was called by us.
+        dt = sample.metrics.time - self.last_sample
+        if dt < 0:
+            dt = 0
+        frames = [(frame.filename, frame.line, frame.function) for frame in sample.frames]
+
+        # Find the depth in the stack that changed between the last sample
+        depth = None  # Default to stacks being identical
+        for i, frame in enumerate(frames):
+            try:
+                if frames[i] != self.last_stack[i]:
+                    depth = i
+                    break
+            except IndexError:
+                depth = i
+                break
+
+        for i, fn in enumerate(frames):
```

Comment:
I'm not familiar with the `pstats` format, but I feel like it should be possible to construct and update the call graph from each sample, with no knowledge of previous samples required. E.g.

~~~ python
if not frames:
    # handle empty frames
    pass

for callee, caller in zip(frames, [{}] + [{c: 1}  for c in frames[:-1]]):
    self.stats.stats[callee] = add_func_stats(
        self.stats.stats.get(callee, (1, 1, 0, 0, {})),
        (caller and 1 or 0, caller and 1 or 0, dt, dt, caller),
    )
~~~

Also, i suspect that `tt` needs to be `0` unless we're at the top of the stack. But again, this could just be me not really knowing the `pstats` format.